File: 백준/Gold/9663. N－Queen/N－Queen.js
function solution(input) {
  const N = parseInt(input, 10);

  const answer = search(
    N,
    0,
    Array(N).fill(false),
    Array(N * 2).fill(false),
    Array(N * 2).fill(false)
  );

  return answer;
}
function search(N, y, width, diagonal1, diagonal2) {
  let count = 0;

  if (N === y) count++;
  else {
    for (let i = 0; i < N; i++) {
      if (width[i] || diagonal1[i + y] || diagonal2[i - y + N]) continue;

      width[i] = diagonal1[i + y] = diagonal2[i - y + N] = true;

      count += search(N, y + 1, width, diagonal1, diagonal2);

      width[i] = diagonal1[i + y] = diagonal2[i - y + N] = false;
    }
  }

  return count;
}

const input = [];
require("readline")
  .createInterface({
    input: process.stdin,
    output: process.stdout,
  })
  .on("line", (line) => {
    input.push(line);
  })
  .on("close", () => {
    console.log(solution(input));
    process.exit();
  });

File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys
input = sys.stdin.readline

def solution():
  n = int(input())

  answer = search(n, 0, [False] * n, [False] * ( n * 2), [False] * ( n * 2))

  return answer

def search(n, y, width, diagonal1, diagonal2):
  count = 0

  if n == y:
    return 1
  
  for i in range(n):
    if width[i] or diagonal1[i+y] or diagonal2[i-y+n]:
      continue
    
    width[i] = diagonal1[i+y] = diagonal2[i-y+n] = True
    count += search(n, y+1, width, diagonal1, diagonal2)
    width[i] = diagonal1[i+y] = diagonal2[i-y+n] = False

  return count
  
print(solution())